src/Symfony/Component/HttpClient/RetryableHttpClient.php

```diff
@@ -53,41 +54,73 @@ public function __construct(HttpClientInterface $client, RetryDeciderInterface $
 
     public function request(string $method, string $url, array $options = []): ResponseInterface
     {
+        if ($this->maxRetries <= 0) {
+            return $this->client->request($method, $url, $options);
+        }
+
         $retryCount = 0;
+        $content = '';
+        $firstChunk = null;
 
-        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount) {
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount, &$content, &$firstChunk) {
             $exception = null;
             try {
                 if ($chunk->isTimeout() || null !== $chunk->getInformationalStatus()) {
                     yield $chunk;
 
                     return;
                 }
-
-                // only retry first chunk
-                if (!$chunk->isFirst()) {
-                    $context->passthru();
-                    yield $chunk;
-
-                    return;
-                }
             } catch (TransportExceptionInterface $exception) {
                 // catch TransportExceptionInterface to send it to strategy.
             }
 
             $statusCode = $context->getStatusCode();
             $headers = $context->getHeaders();
-            if ($retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $partialResponse = new MockResponse($context->getContent(), ['http_code' => $statusCode, 'headers' => $headers]), $exception)) {
-                $context->passthru();
-                yield $chunk;
-
-                return;
+            if (null === $exception) {
+                if ($chunk instanceof FirstChunk) {
+                    $shouldRetry = $this->decider->shouldRetry($method, $url, $options, $statusCode, $headers, null);
+
+                    if (false === $shouldRetry) {
+                        $context->passthru();
+                        yield $chunk;
+
+                        return;
+                    }
+
+                    // Decider need body to decide
+                    if (null === $shouldRetry) {
+                        $firstChunk = $chunk;
+                        $content = '';
+                        // disable exception thrown on destruct
+                        $context->getResponse()->getStatusCode();
```

Comment:
can you apply this patch instead of these lines (and confirm it works):
```diff
--- a/src/Symfony/Component/HttpClient/Response/AsyncResponse.php
+++ b/src/Symfony/Component/HttpClient/Response/AsyncResponse.php
@@ -288,6 +288,9 @@ final class AsyncResponse implements ResponseInterface, StreamableInterface
             if (null !== $chunk->getError()) {
                 // no-op
             } elseif ($chunk->isFirst()) {
+                // Ensure non exception is thrown on destruct for the wrapped response
+                $r->response->getStatusCode();
+
                 $e = $r->openBuffer();
 
                 yield $r => $chunk;
```